### PR TITLE
x/incentive: genesis validation

### PR DESCRIPTION
**Pull request by fedekunze**

### CODE REVIEW

File: x/incentive/types/genesis.go [hunk 1]
```diff
@@ -12,9 +15,35 @@ type GenesisClaimPeriodID struct {
...
+
 // GenesisClaimPeriodIDs array of GenesisClaimPeriodID
 type GenesisClaimPeriodIDs []GenesisClaimPeriodID
 
+// Validate checks if all the GenesisClaimPeriodIDs are valid and there are no duplicated
+// entries.
+func (gcps GenesisClaimPeriodIDs) Validate() error {
+	seenIDS := make(map[string]bool)
+	for _, gcp := range gcps {
+		if seenIDS[gcp.Denom+string(gcp.ID)] {
```
Comment (alexanderbez) [inline x/incentive/types/genesis.go:34]: nit: `gcp.Denom+string(gcp.ID)` is used twice, so I'd recommend storing this into a variable.

File: x/incentive/types/params.go [hunk 2]
```diff
@@ -154,9 +124,47 @@ func (r Reward) String() string {
...
+// Validate performs a basic check of a reward fields.
+func (r Reward) Validate() error {
+	if !r.AvailableRewards.IsValid() {
+		return fmt.Errorf("invalid reward coins %s for %s", r.AvailableRewards, r.Denom)
+	}
+	if !r.AvailableRewards.IsPositive() {
+		return fmt.Errorf("reward amount must be positive, is %s for %s", r.AvailableRewards, r.Denom)
+	}
+	if r.Duration <= 0 {
+		return fmt.Errorf("reward duration must be positive, is %s for %s", r.Duration.String(), r.Denom)
```
Comment (alexanderbez) [inline x/incentive/types/params.go:136]: nit: you don't need to use `String()` when you pass the `%s` format

File: x/incentive/types/rewards.go [hunk 3]
```diff
@@ -72,9 +104,51 @@ func NewClaimPeriod(denom string, id uint64, end time.Time, timeLock time.Durati
...
+	}
+	if cp.TimeLock == 0 {
+		return errors.New("claim period time lock cannot be 0")
+	}
+	return sdk.ValidateDenom(cp.Denom)
+}
+
+// String implements fmt.Stringer
+func (cp ClaimPeriod) String() string {
+	return fmt.Sprintf(`Claim Period:
```
Comment (alexanderbez) [inline x/incentive/types/rewards.go:123]: You can (should?) just use YAML here to make this cleaner.

File: x/incentive/types/rewards.go [hunk 4]
```diff
@@ -72,9 +104,51 @@ func NewClaimPeriod(denom string, id uint64, end time.Time, timeLock time.Durati
...
+
 // ClaimPeriods array of ClaimPeriod
 type ClaimPeriods []ClaimPeriod
 
+// Validate checks if all the ClaimPeriods are valid and there are no duplicated
+// entries.
+func (cps ClaimPeriods) Validate() error {
+	seenPeriods := make(map[string]bool)
+	for _, cp := range cps {
+		if seenPeriods[cp.Denom+string(cp.ID)] {
```
Comment (alexanderbez) [inline x/incentive/types/rewards.go:139]: ditto

File: x/incentive/types/rewards.go [hunk 5]
```diff
@@ -106,6 +194,24 @@ func (c Claim) String() string {
 // Claims array of Claim
 type Claims []Claim
 
+// Validate checks if all the claims are valid and there are no duplicated
+// entries.
+func (cs Claims) Validate() error {
+	seemClaims := make(map[string]bool)
+	for _, c := range cs {
+		if c.Owner != nil && seemClaims[c.Denom+string(c.ClaimPeriodID)+c.Owner.String()] {
```
Comment (alexanderbez) [inline x/incentive/types/rewards.go:202]: ditto

File: x/incentive/types/rewards.go (hunk 3 above)
Comment (fedekunze) [reply x/incentive/types/rewards.go:123]: Agree, there's an issue for that #492

Comment (alexanderbez) [reply x/incentive/types/rewards.go:123]: But it's a code change you're adding here. It'd be one thing of the code already existed.

Comment (fedekunze) [reply x/incentive/types/rewards.go:123]: I just moved it. But I can update it to use yaml 👍

**Pull request merged**